### PR TITLE
build: wire ASan+UBSan sanitizer preset and CI job (#237)

### LINKED ISSUE #237: build: "debug" preset is documented as running sanitizers but no -fsanitize is wired (ASan/UBSan never run in CI)
## Summary

The `debug` build is described in the docs as the sanitizer build, but **no sanitizer flags are present anywhere** in the CMake configuration. As a result, memory-safety and UB bugs are invisible to `ctest`/CI. This is the reason the DICOM CT heap-buffer-overflow (companion issue) can sit in the tree undetected: the debug preset compiles it, runs it, and reports success.

## Evidence

- `DEVELOPER.md:658` (preset table) lists: `debug | build_debug/ | -O0 -g | Day-to-day development, sanitizers`.
- `CLAUDE.md` / `llms.txt` describe the debug preset as "unoptimised, sanitiser-friendly" / "debug + sanitisers".
- Actual flags: `CMakeLists.txt:66-78` adds only, for `Debug`:
  ```
  -Og -g3 -fno-omit-frame-pointer
  ```
  (Note it is `-Og`, not the documented `-O0`.)
- `grep -rn "fsanitize\|SANITIZE" CMakeLists.txt CMakePresets.json` returns **nothing**. `CMakePresets.json` defines only configure/build presets with no sanitizer cache variables.

So neither AddressSanitizer nor UndefinedBehaviorSanitizer is ever enabled by any preset, and CI (which builds `debug` and runs `ctest`) never exercises them.

## Impact

- Whole classes of bugs (heap OOB, use-after-free, UB) pass CI silently. The DICOM CT over-read in the companion issue was only found by compiling that translation unit *manually* with `-fsanitize=address`; the standard `cmake --preset debug && ctest` run reports all-green on the same code.
- The documentation overstates the guarantees of the debug build, so contributors reasonably assume ASan coverage they are not getting.

## Suggested fix

Add an opt-in sanitizer to the debug build so CI can run an instrumented pass. Two common shapes, either is fine:

**Option A — a cache option gating sanitizer flags (works with existing presets):**
```cmake
option(OSH_ENABLE_SANITIZERS "Enable ASan+UBSan on Debug builds" OFF)
if(OSH_ENABLE_SANITIZERS AND CMAKE_C_COMPILER_ID MATCHES "GNU|Clang")
    add_compile_options(-fsanitize=address,undefined -fno-sanitize-recover=all)
    add_link_options(-fsanitize=address,undefined)
endif()
```
Then either flip it on in the `debug` preset's `cacheVariables` (`"OSH_ENABLE_SANITIZERS": "ON"`), or add a dedicated `asan` configure/build preset in `CMakePresets.json` that sets it, and run that preset as an extra CI job.

**Option B — a dedicated `asan` preset** in `CMakePresets.json` (`binaryDir: build_asan`) that sets `CMAKE_C_FLAGS`/`CMAKE_EXE_LINKER_FLAGS` to `-fsanitize=address,undefined -g -Og`, added as a CI matrix entry alongside the normal debug build.

Notes:
- MSVC has `/fsanitize=address` (ASan) but not UBSan; gate the flags on `GNU|Clang` (as above) and, if desired, add `/fsanitize=address` for MSVC separately. This keeps the Windows target building.
- Run the existing `ctest` suite under the sanitized build. Pair this with the DICOM CT regression test so that specific overflow is caught going forward.
- Once wired, correct the `debug` row in `DEVELOPER.md:658` (and the `CLAUDE.md`/`llms.txt` wording) to match reality: either "debug = `-Og -g3`, no sanitizer" plus a separate "asan = sanitizer build", or document that the debug preset now enables sanitizers.

## Acceptance

A CI job builds with `-fsanitize=address[,undefined]` and runs `ctest` green; the DICOM CT non-uniform-slice case (companion issue) fails that job until fixed. Docs describing the debug preset match the flags it actually sets.

### PR DESCRIPTION
**Pull request by grzanka**
## Why

Closes #237.

The `debug` preset was documented as the sanitizer build (DEVELOPER.md preset table, CLAUDE.md, llms.txt), but `grep -rn fsanitize` returned nothing — no `-fsanitize` flag was set anywhere. Memory-safety and UB bugs therefore passed `ctest`/CI silently on the same code that only broke when a translation unit was compiled *manually* with `-fsanitize=address` (see the companion DICOM CT overflow, #235). The docs also overstated the debug build's guarantees.

## What

- **CMake:** add an opt-in `OSH_ENABLE_SANITIZERS` cache option — `-fsanitize=address,undefined -fno-sanitize-recover=all` on GCC/Clang, `/fsanitize=address` on MSVC (which has no UBSan). Default `OFF`, so everyday debug builds and the Windows target are unaffected.
- **Preset:** add a dedicated `asan` configure/build preset (`build_asan/`, Debug + `OSH_ENABLE_SANITIZERS=ON`) so the flags are opt-in per build rather than slowing the everyday `debug` preset.
- **CI:** add a `sanitizers` job to `test.yml` that builds the `asan` preset and runs the full `ctest` suite instrumented. `-fno-sanitize-recover=all` + `halt_on_error`/`abort_on_error` make any sanitizer report a non-zero exit so the run actually fails.
- **Docs:** correct the `debug` row (it is `-Og -g3`, not `-O0`, and has no sanitizer) in DEVELOPER.md, CLAUDE.md and llms.txt, and document the new `asan` preset.

## Test-code bugs surfaced by the new job (fixed here so it is green)

Running the suite under ASan turned up two genuine stack bugs — exactly the class this job exists to catch:

- `test_osh_scoring_parse_geometry`: `tokenize()` stores pointers *into* an `l[]` buffer, but the buffer went out of scope before the parse call read `words[]` → **stack-use-after-scope**. Fixed by keeping each call in the block that owns its buffer.
- `test_osh_scoring_step`: a 1-element dummy stopping-power table with `nenergy = 0` made `osh_material_runtime_sp_lookup()` read one element past the end → **stack-buffer-overflow**. Fixed with a valid 2-point log-energy grid. (The looked-up value only feeds `sp_tr`, unused by the Qeff pages, so the assertions are unchanged.)

## Notes

- **LeakSanitizer is disabled** in the job (`ASAN_OPTIONS=detect_leaks=0`). The tree still has benign at-exit "leaks" (memory reachable at process teardown, e.g. the gemca zone-compile AST) that would otherwise mask the memory-safety signal. Leak cleanup is a separate, larger effort; the job can flip leak detection on afterwards.
- The **companion DICOM CT overflow (#235)** has no test in the suite yet, so this job is green today; adding that regression test in #235 will make it fail under ASan until the overflow is fixed — the intended behavior per #237's acceptance.

## Validation

Locally, `cmake --preset asan && cmake --build --preset asan` then `ASAN_OPTIONS=detect_leaks=0 ctest --test-dir build_asan` is green (GCC 13.3). The plain `debug` suite is also green. (The one local failure, `version_components_in_string`, is an artifact of a tag-less sandbox checkout — `git describe` returns a bare hash — and fails identically on the unmodified debug build; real CI uses `fetch-depth: 0`.)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01TuSePsxbMj8JeT5udd1PLY

---
_Generated by [Claude Code](https://claude.ai/code/session_01TuSePsxbMj8JeT5udd1PLY)_

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -105,6 +105,25 @@ if(OSH_ENABLE_COVERAGE)
...
+# -fno-sanitize-recover=all turns UBSan reports into hard aborts so a violation
+# fails the test run instead of merely printing a warning.
+option(OSH_ENABLE_SANITIZERS "Enable AddressSanitizer + UndefinedBehaviorSanitizer" OFF)
+if(OSH_ENABLE_SANITIZERS)
+    if(CMAKE_C_COMPILER_ID MATCHES "GNU|Clang")
+        add_compile_options(-fsanitize=address,undefined -fno-sanitize-recover=all -fno-omit-frame-pointer)
+        add_link_options(-fsanitize=address,undefined)
+    elseif(MSVC)
+        add_compile_options(/fsanitize=address)
+    else()
```
[the remark this answers is not in the document]
Comment (grzanka) [reply CMakeLists.txt:131]: Addressed in 7ab8456. The MSVC branch now strips `/RTC*` from the Debug C flags — those run-time checks are incompatible with `/fsanitize=address` and `cl.exe` errors out otherwise — and forces `/INCREMENTAL:NO`, since ASan cannot use incremental linking. The ASan runtime is import-linked automatically by `cl.exe`; its `clang_rt.asan_dynamic-*.dll` must be discoverable on `PATH` when the instrumented tests run.

Caveat: there is still no Windows sanitizer CI job, so this MSVC path is not yet validated end-to-end (it can't be exercised on the Linux/macOS legs). Standing up a `windows-latest` ASan leg and confirming a green run is tracked in #240.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by nbassler — APPROVED** [review]
